### ISSUE TITLE
participate in google summer of code as nixos foundation

### ISSUE THREAD

**Issue by Janik-Haag** [NixOS/foundation#115]
Google opened there applications for summer of code last week at https://summerofcode.withgoogle.com/ there deadline is until ***06.02.2024***.

tldr:
gsoc has a mentee/mentor model where mentors give some tasks to there mentee/s and they should meet regularly.
google pays mentees some compensation. 
for more info please refer to the gsoc homepage.

it would be neat to have nixpkgs and maybe nix participate, depending on scope/time/motivation some things that could be done in gsoc are:
- writing packages/modules for nixpkgs
- doing larger structural changes/refactoring existing modules
- working on some rfc implementation

don't really have ideas about nix since I'm personally only really involved with nixpkgs

From my understanding the application is quite straight forward and the only to conditions are:

- having a foss license listed [here](https://opensource.org/licenses/) which is given for both nix and nixpkgs
- and accepting this [terms](https://summerofcode.withgoogle.com/terms/org)

I would be willing to do the application for the nixos org and write a discourse post looking for mentors, if the foundation gives me a okay, so I have to say that the gsoc form says:
> This should only be completed by an official representative of an organization (only 1 registration per org)

**Comment by zimbatm** [comment]
Great initiative. Based on previous years, I don't think there are any objections on the Foundation side. 

The hardest part is getting organized in time.

https://summerofcode.withgoogle.com/rules is where the requirements are listed.

Do you mind organizing a call for mentors with a post on Discourse?

The biggest mistake we made in the previous year was to have everybody contribute to the list of projects. Everybody is here to propose work to do, but then nobody wants to mentor. So this year, I propose we do the following instead: only mentors can propose projects they are interested in mentoring. If you want something to be worked on, then become a mentor.

**Comment by Janik-Haag** [comment]
> The biggest mistake we made in the previous year was to have everybody contribute to the list of projects. Everybody is here to propose work to do, but then nobody wants to mentor. So this year, I propose we do the following instead: only mentors can propose projects they are interested in mentoring. If you want something to be worked on, then become a mentor.

Yeah sounds reasonable to look for mentors and not projects, esp. since most people mentoring should be experienced enough to be aware of some of the nixpkgs todos and if not finding projects is still easier then finding mentors.
> Do you mind organizing a call for mentors with a post on Discourse?

I will write a `call for gsoc mentors` draft later.

**Comment by Janik-Haag** [comment]
Here is a draft discourse post, would be awesome if you find time to review it and tell me if I can post it. 

---

# call for gsoc mentors

Hey Everyone,

I'm glad to announce that we want to participate in this years Google Summer of Code with nix/nixpkgs 🎉
This means we are now looking for experienced community members keen on sharing there knowledge with open source newcomers and mentoring them.

As a gsoc mentor you would be expected to take a few hours out of your week to meet and chat with your potential mentee, this will probably be more time at the start and should progressively get less over the summer. In those meetings you will introduce the mentee to our community and help them work on a predefined Project.

We have to set two hard requirements for mentors:
- being capable of fluently speaking English
- having at least a age of 18 year's or older

You might ask what's in it for you:
Mentoring someone is not just a great learning experience for them but also for yourself, you get to improve your own teaching ability which is a great skill in almost any life situation while also refreshing your self on all the nix fundamentals.
And being able to say you participated as a mentor in a program like gsoc also looks great on any cv ;)

Please visit the [gsoc-website](https://summerofcode.withgoogle.com/) for more information about Google Summer of Code.
If you want to keep in the loop on how far we are with gsoc applications, etc.. you can subscribe to this [issue](https://github.com/NixOS/foundation/issues/115)

I hope this post sparked your interest, please leave a quick message here if you are interested in becoming a gsoc mentor or message me on matrix (@janik0:matrix.org) or via email (gsoc.janik@aq0.de) and read the [official mentor guide](https://google.github.io/gsocguides/mentor).

**Comment by zimbatm** [comment]
Love it. Let's go!

Another benefit is that mentors can help push projects they are interested in seeing happening. However, with students, the results aren't always guaranteed.

**Comment by Janik-Haag** [comment]
Okay so next step would be filing the org registration [here](https://summerofcode.withgoogle.com/register/organization), I can do so if I get the foundations greenlight, otherwise it would be awesome if one of you could do that. I think we should file it as soon as possible to avoid getting time pressure and because we need to do so to afterwards register mentors.

**Comment by Janik-Haag** [comment]
@thufschmitt created the gsoc nixos org and invited me, I'll fill out the `Organization Questionnaire` and ` Organization Profile` and ask if there is any thing I can't answer. Thanks for getting this moving so fast :D

**Comment by GetPsyched** [comment]
Great initiative, love to see things moving fast.

Was curious regarding how this would overlap with [ngi-nix/summer-of-nix](https://github.com/ngi-nix/summer-of-nix). Would these programmes stay separate from one another?

**Comment by Janik-Haag** [comment]
@GetPsyched there will be a Summer of Nix 2024 which will be independent of GSoC. For more info please move the discussion to discourse or matrix since it's somewhat offtopic for this issue, the right person to ask here is probably @fricklerhandwerk

**Comment by Janik-Haag** [comment]
We need a `Ideas List` and `Contributor Guidance` web page for GSoC, I would suggest we create a gsoc repo in the nixos or nix-community github-org and just put some markdown there and maybe render it with github-pages and just let mentors PR their ideas, wdyt?

Ideas List:
> Prospective GSoC Contributors will view this link to understand the kinds of projects available in your organization. A clean and simple presentation is best. Avoid links to unfiltered bugtrackers or other specialized tools. 

Contributor Guidance:
>  Provide your potential contributors with a page containing tips on how to write a successful proposal for your organization. Let them know what you want included, how you want it structured, and how to best get in touch. Examples.

**Comment by zimbatm** [comment]
@Janik-Haag, since it's an "official" project, I created a team and the repo in the official org. I also started a skeleton README, but feel free to change anything. https://github.com/NixOS/GSoC2024/

**Comment by fricklerhandwerk** [comment]
@Zimbatm I strongly recommend renaming that repo to just GSoC, since there's a good chance this will continue next year. I made very bad experience with uncontrolled proliferation of places in the context of Summer of Nix. It's a mess to navigate, and incurs downstream costs of mitigation.

@GetPsyched Summer of Nix is a completely independent thing funded through NGI, and is merely modeled after GSoC to some extent. As @Janik-Haag wrote, it will continue this year. Part of the preparation is cleaning up the sprawl of repositories created in the past years.

**Comment by GetPsyched** [comment]
> @GetPsyched Summer of Nix is a completely independent thing funded through NGI, and is merely modeled after GSoC to some extent. As @Janik-Haag wrote, it will continue this year. Part of the preparation is cleaning up the sprawl of repositories created in the past years.

Thanks for the clarification, @fricklerhandwerk and @Janik-Haag. Apologies for posting the query here.

Looking forward to contributing in GSoC with the Foundation!

**Comment by eldios** [comment]
Hi, just getting my feet wet in contributing to the nix-ecosystem and I'll be happy to contribute.
I already contacted @Janik-Haag via Matrix but you can expect more contributions in the coming days/weeks/months as I dive deeper in the ecosystem.

that said..

> @zimbatm I strongly recommend renaming that repo to just GSoC, since there's a good chance this will continue next year. I made very bad experience with uncontrolled proliferation of places in the context of Summer of Nix. It's a mess to navigate, and incurs downstream costs of mitigation.

☝🏻 definitely this.. and then maybe put a section, a directory or some subsection that points to the current year. 
That way it's as easy as copy-pasting or create a template to replicate each year.

More of my observation and contribution will be made on that repo.

Cheers everybody 🖖🏻

**Comment by harsh-ps-2003** [comment]
Does this mean there is not going to be a Summer of Nix 2024 :frowning: The problem with Google Summer of Code is that if you are pretty experienced with Open Source stuff, they don’t allow you to participate, even if you are a student. Its only for people very new to Open Source software development. Now, I was a Google Summer of Code student, but now as I maintain software and also going to be a mentor for my organization, they wont let me participate most probably which is a total bummer! I was dreaming of experiencing mob programming with Nix community but its all shattered now! @Janik-Haag

**Comment by Janik-Haag** [comment]
Basically everything from my side of things is done now.

- All the Questionnaire questions are answered
- The GSoC NixOS org profile has a 100% completion 
- We have a [contributor guidance document](https://github.com/NixOS/GSoC2024/blob/main/contributor-guidance.md)
- And a [idea list](https://github.com/NixOS/GSoC2024/pull/3/)
- There is matrix rooms ([gsoc public](https://matrix.to/#/#gsoc-nix:matrix.org), mentors invite only, organizers invite only)

There is a few more things we can do now but they are all optional except for two:

We need to get at least four solid ideas in the ideas list. I'm relying on the people who said they would mentor to PR more, if we don't meet that quota we just get disqualified.
Someone needs to press  on the submit button in the gsoc org, I will do that on monday evening and just hope for the best.
I did all the things on best effort and am sure a lot of them can be improved and there might be some typos, so if someone volunteers to review the application before I send it that would be awesome.
We should still rename github.com/nixos/gsoc2024 to github.com/nixos/gsoc but I can't do that because of permissions.
We might also want to do some github pages rendering of the idea list but  that has to be enabled in the repo settings.
And then we might also want to have some stable urls for both idea list and contributor guidance since we have to link them in the application, but I can't do that either. 

Enjoy your weekend and have a nice event to everyone attending fosdem - I'm gonna try to take a break for today.

**Comment by Janik-Haag** [comment]
Application is done, now we just need to wait and see if we get accepted

**Comment by refroni** [comment]
Adding that I've reached out to a few contacts at Google, will update

**Comment by Janik-Haag** [comment]
We got accepted :tada: 

![image](https://github.com/NixOS/foundation/assets/80165193/e9c78d14-ef58-43c8-b8ec-730285cb2bf9)

**Comment by Janik-Haag** [comment]
I think we can close this issue since we got accepted :tada: 
And the other organization can be done in the gsoc repo.
Thank you everyone for helping!

https://chaos.social/@nixos_org/111970809461688895
